subsys/bluetooth/host/adv.c

```diff
@@ -27,6 +27,104 @@ enum adv_name_type {
 	ADV_NAME_TYPE_SD,
 };
 
+struct bt_ad {
+	const struct bt_data *data;
+	size_t len;
+};
+
+struct ad_stream {
+	/* ad is a two dimensional array of struct bt_data elements. */
+	const struct bt_ad *ad;
+	size_t ad_len;
+
+	/* Indices into the two dimensional array of elements. */
+	size_t ad_index;
+	size_t data_index;
+
+	/* Current LTV offset contains the offset in the AD structure.
+	 * The length and type are included in this offset.
+	 */
+	uint16_t current_ltv_offset;
+};
+
+static void ad_stream_new(struct ad_stream *stream, const struct bt_ad *ad, size_t ad_len)
+{
+	(void)memset(stream, 0, sizeof(*stream));
+	stream->ad = ad;
+	stream->ad_len = ad_len;
+}
+
+static bool ad_stream_is_empty(const struct ad_stream *stream)
+{
+	return stream->ad_index == stream->ad_len;
+}
+
+static const struct bt_data *ad_stream_current_ltv_get(struct ad_stream *stream)
+{
+	const struct bt_data *current_ltv = &stream->ad[stream->ad_index].data[stream->data_index];
+	const bool current_ltv_complete = (stream->current_ltv_offset == current_ltv->data_len + 2);
+
+	if (current_ltv_complete) {
+		stream->current_ltv_offset = 0;
+
+		if (stream->data_index + 1 == stream->ad[stream->ad_index].len) {
+			stream->data_index = 0;
+			stream->ad_index++;
```

Comment:
This function is called `current_ltv_get`, but we increase the `ad_index` here, and return `stream->ad[stream->ad_index]` 
later. 

That means that it does not return the current, but may return either the current or the next, while also modifying the input, which should (almost) never be done for a `get` function.

Reply:
What about naming it `ad_stream_current_ltv_update()`, or do you have a proposal for what and how to change it? I can simply copy out the content of this extracted function and paste it in the `_read`. I don't think that makes the code more readable.

Reply:
`ad_stream_current_ltv_update` sounds like a decent solution; as least the caller would expect the input to change then.